In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report


---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# YOUR CODE HERE!
df=pd.read_csv("Resources/lending_data.csv")
# Review the DataFrame
# YOUR CODE HERE!
df.sample(5)

loan_size  interest_rate  borrower_income  debt_to_income  \
2063      9400.0          7.141            47800        0.372385   
44647     7900.0          6.501            41800        0.282297   
7515      9000.0          6.942            45900        0.346405   
29391     8200.0          6.606            42800        0.299065   
2079     11100.0          7.856            54500        0.449541   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
2063                 4                 0       17800            0  
44647                2                 0       11800            0  
7515                 3                 0       15900            0  
29391                2                 0       12800            0  
2079                 5                 1       24500            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
# YOUR CODE HERE!]
y = df['loan_status']
# Separate the X variable, the features
# YOUR CODE HERE!
X =df.drop(columns=['loan_status'])

In [4]:
# Review the y variable Series
# YOUR CODE HERE!
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
# YOUR CODE HERE!
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
# YOUR CODE HERE!
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, random_state=1 ) 


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
model = LogisticRegression(solver='lbfgs', max_iter = 200, random_state=1)
# Fit the model using training data
# YOUR CODE HERE!
model.fit(X_train, y_train)

LogisticRegression(max_iter=200, random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
# YOUR CODE HERE!
y_pred = model.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [10]:
# Print the balanced_accuracy score of the model
# YOUR CODE HERE!
balanced_accuracy_score(y_test, y_pred)

0.9442676901753825

In [11]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
confusion_matrix(y_test, y_pred)

array([[18679,    80],
       [   67,   558]], dtype=int64)

In [12]:
# Print the classification report for the model
# YOUR CODE HERE!
target_names = ["healthy loan", "high-risk loan"]
print(classification_report(y_test, y_pred, target_names=target_names))

                precision    recall  f1-score   support

  healthy loan       1.00      1.00      1.00     18759
high-risk loan       0.87      0.89      0.88       625

      accuracy                           0.99     19384
     macro avg       0.94      0.94      0.94     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:**  The accuracy score is 99%, indicating exceptional overall performance of the model. For the "healthy loan" class, the precision is perfect, indicating that all predicted healthy loans are correct. For the "high-risk loan" class, the precision is 87%, meaning that 87% of the predicted high-risk loans are correct. Recall, also known as sensitivity, measures the model's ability to capture all instances of a specific label. The recall for the "healthy loan" class is perfect, but for the "high-risk loan" class is 89%, meaning that the model identifies only 89% of the high-risk loans.
One area that requires improvement is the reduction of False Rejections, which is currently at 67. This reduction is crucial to minimize missed opportunities. In scenarios like identifying high-risk loans or detecting fraud, false negatives can result in missed chances. By minimizing false negatives, we significantly enhance our ability to accurately identify positive instances, enabling us to take appropriate actions and seize those valuable opportunities.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [13]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
ROS_model = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
# YOUR CODE HERE!
ROS_X, ROS_y= ROS_model.fit_resample(X_train, y_train)

In [14]:
# Count the distinct values of the resampled labels data
# YOUR CODE HERE!
ROS_y.value_counts()

0    56277
1    56277
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [15]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
model = LogisticRegression(solver='lbfgs', random_state=1)
# Fit the model using the resampled training data
# YOUR CODE HERE!
model.fit(ROS_X, ROS_y)
# Make a prediction using the testing data
# YOUR CODE HERE!
ROS_pred = model.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [16]:
# Print the balanced_accuracy score of the model 
# YOUR CODE HERE!
balanced_accuracy_score(y_test, ROS_pred)

0.9959744975744975

In [17]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
confusion_matrix(y_test, ROS_pred)

array([[18668,    91],
       [    2,   623]], dtype=int64)

In [18]:
# Print the classification report for the model
# YOUR CODE HERE!
target_names = ["healthy loan", "high-risk loan"]
print(classification_report(y_test, ROS_pred, target_names=target_names))

                precision    recall  f1-score   support

  healthy loan       1.00      1.00      1.00     18759
high-risk loan       0.87      1.00      0.93       625

      accuracy                           1.00     19384
     macro avg       0.94      1.00      0.96     19384
  weighted avg       1.00      1.00      1.00     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model fit with oversampled data outperforms the model fit with the original data in multiple aspects. The precision of the model for the "healthy loan" class is 100%, while for the "high-risk loan" class, it is 87%. The recall is 100% for both classes, indicating that the model captures all actual instances. The F1-score is 100% for the "healthy loan" class and 0.93 for the "high-risk loan" class, demonstrating a good balance between precision and recall. The accuracy score is 100%, indicating exceptional overall performance. By significantly reducing the number of false negatives from 67 to 2, the model enhances the identification of positive instances, allowing for appropriate actions to be taken. Minimizing false negatives optimizes resource allocation and fosters trust in the model's reliability and credibility. Considering these factors, the logistic regression model trained with oversampled data emerges as the more favorable choice compared to the model trained with the original data.